### PR TITLE
Implement SSE3 support in V86

### PR DESCRIPTION
**Pull request by spetterman66**

### CODE REVIEW

File: gen/x86_table.js [hunk 1]
```diff
@@ -621,15 +621,19 @@ const encodings = [
     { sse: 1, opcode: 0x0F10, e: 1, custom: 1 },
     { sse: 1, opcode: 0xF30F10, e: 1, custom: 1 },
     { sse: 1, opcode: 0x660F10, e: 1, custom: 1 },
+    { sse: 1, opcode: 0x660F7C, e: 1, custom: 1 }, // sse3
+    { sse: 1, opcode: 0x660F7D, e: 1, custom: 1 }, // sse3
     { sse: 1, opcode: 0xF20F10, e: 1, custom: 1 },
+    { sse: 1, opcode: 0xF20F7C, e: 1, custom: 1 }, // sse3
```
Comment (copy) [inline gen/x86_table.js:627]: This one is already defined (see the CI failure)

Comment (spetterman66) [reply gen/x86_table.js]: it passed make all-debug :D

**Comment by spetterman66** [comment]
if this doesnt work then i will blow up

**Comment by spetterman66** [comment]
if this aint working then i will blow up

**Comment by spetterman66** [comment]
interesting :-/

**Comment by spetterman66** [comment]
i cannot for the life of me figure out how to solve this 😕

File: gen/x86_table.js [hunk 2]
```diff
@@ -324,12 +324,12 @@ const encodings = [
...
     { opcode: 0xDD, e: 1, fixed_g: 0, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, },
-    { opcode: 0xDD, e: 1, fixed_g: 1, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, skip_mem: 1, }, // unimplemented: fisttp (sse3)
+    { opcode: 0xDD, e: 1, fixed_g: 1, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, }, // fisttp (sse3)
     { opcode: 0xDD, e: 1, fixed_g: 2, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, },
     { opcode: 0xDD, e: 1, fixed_g: 3, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, },
-    { opcode: 0xDD, e: 1, fixed_g: 4, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, skip_mem: 1 }, // frstor
+    { opcode: 0xDD, e: 1, fixed_g: 4, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, }, // frstor
     { opcode: 0xDD, e: 1, fixed_g: 5, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, },
-    { opcode: 0xDD, e: 1, fixed_g: 6, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, skip_mem: 1 }, // fsave
+    { opcode: 0xDD, e: 1, fixed_g: 6, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, }, // fsave
```
Comment (copy) [inline gen/x86_table.js:332]: Put the skip_mem back into fsave/frstor (it's unrelated to sse3)

File: gen/x86_table.js [hunk 3]
```diff
@@ -324,12 +324,12 @@ const encodings = [
     { opcode: 0xDC, e: 1, fixed_g: 7, custom: 1, is_fpu: 1, task_switch_test: 1, },
 
     { opcode: 0xDD, e: 1, fixed_g: 0, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, },
-    { opcode: 0xDD, e: 1, fixed_g: 1, custom: 0, is_fpu: 1, task_switch_test: 1, os: 1, skip_mem: 1, }, // unimplemented: fisttp (sse3)
+    { opcode: 0xDD, e: 1, fixed_g: 1, custom: 1, is_fpu: 1, task_switch_test: 1, os: 1, }, // fisttp (sse3)
```
Comment (copy) [inline gen/x86_table.js:327]: Leave `custom: 0` (it's for jitting)

File: gen/x86_table.js [hunk 4]
```diff
@@ -341,14 +341,14 @@ const encodings = [
...
-    { opcode: 0xDF, e: 1, fixed_g: 3, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 4, custom: 1, is_fpu: 1, task_switch_test: 1, skip_mem: 1 }, // unimplemented: Binary Coded Decimals
-    { opcode: 0xDF, e: 1, fixed_g: 5, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 6, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 7, custom: 1, is_fpu: 1, task_switch_test: 1 },
+    { opcode: 0xDF, e: 1, fixed_g: 0, custom: 1, is_fpu: 1, task_switch_test: 1, },
+    { opcode: 0xDF, e: 1, fixed_g: 1, custom: 1, is_fpu: 1, task_switch_test: 1, }, // fisttp (sse3)
+    { opcode: 0xDF, e: 1, fixed_g: 2, custom: 1, is_fpu: 1, task_switch_test: 1, },
+    { opcode: 0xDF, e: 1, fixed_g: 3, custom: 1, is_fpu: 1, task_switch_test: 1, },
+    { opcode: 0xDF, e: 1, fixed_g: 4, custom: 1, is_fpu: 1, task_switch_test: 1, }, // unimplemented: Binary Coded Decimals
```
Comment (copy) [inline gen/x86_table.js:348]: Put the skip_mem back in

File: gen/x86_table.js [hunk 5]
```diff
@@ -341,14 +341,14 @@ const encodings = [
...
-    { opcode: 0xDF, e: 1, fixed_g: 0, custom: 0, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 1, custom: 0, is_fpu: 1, task_switch_test: 1, skip_mem: 1 }, // unimplemented: fisttp (sse3)
-    { opcode: 0xDF, e: 1, fixed_g: 2, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 3, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 4, custom: 1, is_fpu: 1, task_switch_test: 1, skip_mem: 1 }, // unimplemented: Binary Coded Decimals
-    { opcode: 0xDF, e: 1, fixed_g: 5, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 6, custom: 1, is_fpu: 1, task_switch_test: 1 },
-    { opcode: 0xDF, e: 1, fixed_g: 7, custom: 1, is_fpu: 1, task_switch_test: 1 },
+    { opcode: 0xDF, e: 1, fixed_g: 0, custom: 1, is_fpu: 1, task_switch_test: 1, },
+    { opcode: 0xDF, e: 1, fixed_g: 1, custom: 1, is_fpu: 1, task_switch_test: 1, }, // fisttp (sse3)
```
Comment (copy) [inline gen/x86_table.js:345]: Leave `custom: 0` (both lines)

File: src/rust/cpu/instructions_0f.rs [hunk 6]
```diff
@@ -1839,7 +1915,60 @@ pub unsafe fn instr_F20F7C_reg(r1: i32, r2: i32) { instr_F20F7C(read_xmm128s(r1)
...
+            source.f32[0] + source.f32[1],
+            source.f32[2] + source.f32[3],
+        ],
+    };
+    write_xmm_reg128(r, result);
+}
+pub unsafe fn instr_DD_reg(r1: i32, r2: i32) { instr_DD(read_xmm128s(r1), r2); }
+pub unsafe fn instr_DD_mem(addr: i32, r: i32) {
+    instr_DD(return_on_pagefault!(safe_read128s(addr)), r);
+}
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:1953]: These are floating point instructions. They should be in instructions.rs. See for example fpu_fistm16p.

**Comment by spetterman66** [comment]
nice :/

**Comment by spetterman66** [comment]
sooooooo yeah we're stuck on here :-/

**Comment by spetterman66** [comment]
hmm... but why doesn't sse3 show up on `cat /proc/cpuinfo`?

**Comment by copy** [comment]
> hmm... but why doesn't sse3 show up on cat /proc/cpuinfo?

Good question. It does show up in other OSes (e.g. KolibriOS).

**Comment by copy** [comment]
Did you test Windows 8 and/or Android with this?

**Comment by Ryan8085309853409** [comment]
Don't have time rn, will do it later (I use this account on my phone)

**Comment by spetterman66** [comment]
(here i am btw)

**Comment by spetterman66** [comment]
well, now we have sse3 support, but we have problems with APIC i guess
image

**Comment by spetterman66** [comment]
image
android-x86 4.3 without ACPI fails here

**Comment by SuperMaxusa** [comment]
> It does show up in other OSes (e.g. KolibriOS).

Agreed, in CPU-Z detects SSE3, but in Linux's /proc/cpuinfo **pni** (Prescott New Instructions, second name of SSE3) is not detected
cpuz

Also, Alpine Linux 3.18.2 ([iso](https://dl-cdn.alpinelinux.org/alpine/v3.18/releases/x86/alpine-virt-3.18.2-x86.iso)) won't booting for me (on 5664eea5 worked properly):

<details>
<summary>Log</summary>
<pre>
[   62.679000] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[   62.679000]  (detected by 0, t=6002 jiffies, g=-1195, q=1 ncpus=1)
[   62.679000] rcu: All QSes seen, last rcu_preempt kthread activity 6002 (-23924--29926), jiffies_till_next_fqs=1, root ->qsmask 0x0
[   62.679000] rcu: rcu_preempt kthread starved for 6002 jiffies! g-1195 f0x2 RCU_GP_WAIT_FQS(5) ->state=0x0 ->cpu=0
[   62.679000] rcu:     Unless rcu_preempt kthread gets sufficient CPU time, OOM is now expected behavior.
[   62.679000] rcu: RCU grace-period kthread stack dump:
[   62.679000] task:rcu_preempt     state:R  running task     stack:0     pid:15    ppid:2      flags:0x00004000
[   62.679000] Call Trace:
[   62.679000]  __schedule+0x38d/0x1140
[   62.679000]  schedule+0x56/0xf0
[   62.679000]  schedule_timeout+0x6c/0x110
[   62.679000]  ? __bpf_trace_tick_stop+0x30/0x30
[   62.679000]  rcu_gp_fqs_loop+0xfa/0x450
[   62.679000]  ? rcu_gp_init+0x2a2/0x4d0
[   62.679000]  rcu_gp_kthread+0xc5/0x170
[   62.679000]  kthread+0xd5/0x100
[   62.679000]  ? rcu_gp_init+0x4d0/0x4d0
[   62.679000]  ? kthread_complete_and_exit+0x20/0x20
[   62.679000]  ret_from_fork+0x1c/0x28
[   62.679000] rcu: Stack dump where RCU GP kthread last ran:
[   62.679000] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 6.1.51-0-virt #1-Alpine
[   62.679000] EIP: delay_tsc+0x6a/0xd0
[   62.679000] Code: 00 39 d8 75 3b 0f ae e8 0f 31 8b 4d ec 89 c6 89 d7 2b 45 e4 1b 55 e8 39 c8 89 d0 1b 45 f0 73 38 b8 01 00 00 00 e8 66 2e 97 ff <64> a1 34 08 c0 ca 85 c0 75 bc e8 f7 2c 04 00 eb b5 8d 74 26 00 90
[   62.679000] EAX: 00000000 EBX: 00000000 ECX: 00000000 EDX: 00000000
[   62.679000] ESI: c36c4a80 EDI: 00000014 EBP: c10cff08 ESP: c10cfeec
[   62.679000] DS: 007b ES: 007b FS: 00d8 GS: 0000 SS: 0068 EFLAGS: 00000246
[   62.679000] CR0: 80050033 CR2: ff7ff000 CR3: 0ac16000 CR4: 000006b0
[   62.679000] DR0: 00000000 DR1: 00000000 DR2: 00000000 DR3: 00000000
[   62.679000] DR6: 00000000 DR7: 00000000
[   62.679000] Call Trace:
[   62.679000]  <IRQ>
[   62.679000]  ? show_regs.cold+0x14/0x1a
[   62.679000]  ? dump_cpu_task+0x61/0x70
[   62.679000]  ? rcu_check_gp_kthread_starvation.cold+0x157/0x15d
[   62.679000]  ? rcu_sched_clock_irq+0x9a1/0x9d0
[   62.679000]  ? update_process_times+0x63/0x90
[   62.679000]  ? tick_periodic+0x31/0x90
[   62.679000]  ? tick_handle_periodic+0x1e/0x70
[   62.679000]  ? get_stack_info+0x140/0x140
[   62.679000]  ? timer_interrupt+0xf/0x20
[   62.679000]  ? __handle_irq_event_percpu+0x3b/0x160
[   62.679000]  ? handle_irq_event+0x29/0x70
[   62.679000]  ? handle_fasteoi_nmi+0x100/0x100
[   62.679000]  ? handle_level_irq+0x8f/0x170
[   62.679000]  ? __handle_irq+0x9e/0xc0
[   62.679000]  </IRQ>
[   62.679000]  ? __common_interrupt+0x54/0xf0
[   62.679000]  ? common_interrupt+0x34/0x50
[   62.679000]  ? asm_common_interrupt+0x102/0x140
[   62.679000]  ? delay_tsc+0x6a/0xd0
[   62.679000]  ? delay_tsc+0x6a/0xd0
[   62.679000]  __const_udelay+0x2e/0x40
[   62.679000]  test_nmi_ipi.constprop.0+0x94/0xe1
[   62.679000]  local_ipi+0x36/0x52
[   62.679000]  dotest.constprop.0+0xc/0xb6
[   62.679000]  nmi_selftest+0x7e/0x19e
[   62.679000]  native_smp_cpus_done+0x8a/0x191
[   62.679000]  smp_init+0x9f/0xbc
[   62.679000]  kernel_init_freeable+0x13b/0x291
[   62.679000]  ? rest_init+0xc0/0xc0
[   62.679000]  kernel_init+0x12/0x100
[   62.679000]  ret_from_fork+0x1c/0x28
</pre>
</details>

tested commit: https://github.com/copy/v86/pull/906/commits/61231979aa45c6b17a545a8b3d460579c187dc19

**Comment by spetterman66** [comment]
aaaaaaand the ci has failed, looks like some issue with the fpu :/

**Comment by spetterman66** [comment]
then theres smth wrong with the fisttp implementation, idk what else lmao

**Comment by spetterman66** [comment]
so i'm stuck then i guess -_-
[LinkError: WebAssembly.instantiate(): Import #15 module="e" function="fpu_fisttp" error: function import requires a callable]

**Comment by spetterman66** [comment]
I think this CI thing has an error:
```[LinkError: WebAssembly.instantiate(): Import #15 module="e" function="fpu_fisttp" error: function import requires a callable]
/home/runner/work/v86/v86/tests/nasm/run.js:4
process.on("unhandledRejection", exn => { throw exn; });
                                          ^

[LinkError: WebAssembly.instantiate(): Import #15 module="e" function="fpu_fisttp" error: function import requires a callable]

Node.js v18.17.1
Worker error code: 1```

**Comment by copy** [comment]
> I think this CI thing has an error

That's a bug in the code, `fpu_fisttp` isn't defined.

**Comment by spetterman66** [comment]
works now?

**Comment by spetterman66** [comment]
> works now?

yes it does!
image

**Comment by spetterman66** [comment]
we fkn did it 💯 
image

**Comment by spetterman66** [comment]
@copy sorry for being inpatient, but can you merge this pull request, since we have sse3 support?

File: src/rust/cpu/fpu.rs [hunk 7]
```diff
@@ -368,7 +380,13 @@ pub unsafe fn fpu_fistm64p(addr: i32) {
     safe_write64(addr, v as u64).unwrap();
     fpu_pop();
 }
-
+#[no_mangle]
+pub unsafe fn fpu_fisttpm64(addr: i32) {
+    return_on_pagefault!(writable_or_pagefault(addr, 8));
+    let v = fpu_convert_to_i64(fpu_get_st0());
```
Comment (copy) [inline src/rust/cpu/fpu.rs:386]: Use `.trunc()` instead of `fpu_convert_to_i64` (also in the other functions).

File: src/rust/cpu/fpu.rs [hunk 8]
```diff
@@ -341,6 +347,12 @@ pub unsafe fn fpu_convert_to_i32(f: F80) -> i32 {
     x
 }
 #[no_mangle]
+pub unsafe fn fpu_fisttpm32(addr: i32) {
+    return_on_pagefault!(writable_or_pagefault(addr, 4));
+    let v = fpu_convert_to_i32(fpu_get_st0());
+    safe_write32(addr, v).unwrap();
+}
```
Comment (copy) [inline src/rust/cpu/fpu.rs:354]: I believe this is missing an `fpu_pop()` (same for `fpu_fisttpm16`)

File: src/rust/cpu/instructions_0f.rs [hunk 9]
```diff
@@ -658,10 +685,23 @@ pub unsafe fn instr_660F16_mem(addr: i32, r: i32) {
...
-
+pub unsafe fn instr_F30F16(source: reg128, r: i32) {
+    // movshdup xmm, xmm/mem128
+    let destination = read_xmm128s(r);
+    let result = reg128 {
+        f32: [
+            destination.f32[0] + destination.f32[1],
+            destination.f32[2] + destination.f32[3],
+            source.f32[0] + source.f32[1],
+            source.f32[2] + source.f32[3],
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:696]: This implementation isn't correct. See https://www.felixcloutier.com/x86/movshdup

File: src/rust/jit_instructions.rs [hunk 10]
```diff
@@ -5684,7 +5693,26 @@ pub fn instr_F30F10_reg_jit(ctx: &mut JitContext, r1: u32, r2: u32) {
     ctx.builder
         .store_aligned_i32(global_pointers::get_reg_xmm_offset(r2));
 }
-
+pub fn instr_F30F12_mem_jit(ctx: &mut JitContext, modrm_byte: ModrmByte, r: u32) {
+    instr_660F6E_mem_jit(ctx, modrm_byte, r)
```
Comment (copy) [inline src/rust/jit_instructions.rs:5697]: This implementation isn't correct. See https://www.felixcloutier.com/x86/movsldup

File: src/rust/jit_instructions.rs [hunk 11]
```diff
@@ -6134,7 +6162,12 @@ pub fn instr_F20F7C_mem_jit(ctx: &mut JitContext, modrm_byte: ModrmByte, r: u32)
 pub fn instr_F20F7C_reg_jit(ctx: &mut JitContext, r1: u32, r2: u32) {
     sse_read128_xmm_xmm(ctx, "instr_F20F7C", r1, r2);
 }
-
+pub fn instr_F20F7D_mem_jit(ctx: &mut JitContext, modrm_byte: ModrmByte, r: u32) {
+    sse_read128_xmm_mem(ctx, "instr_F20F7D", modrm_byte, r);
+}
+pub fn instr_F20F7D_reg_jit(ctx: &mut JitContext, r1: u32, r2: u32) {
+    sse_read128_xmm_xmm(ctx, "instr_F20F7D", r1, r2);
```
Comment (copy) [inline src/rust/jit_instructions.rs:6169]: https://www.felixcloutier.com/x86/hsubps

File: src/rust/cpu/instructions_0f.rs [hunk 12]
```diff
@@ -554,14 +554,41 @@ pub unsafe fn instr_660F12_mem(addr: i32, r: i32) {
...
-pub unsafe fn instr_F30F12_mem(_addr: i32, _r: i32) { unimplemented_sse(); }
+pub unsafe fn instr_F20F12(source: reg128, r: i32) {
+    // movddup xmm, xmm/mem128
+    let destination = read_xmm128s(r);
+    let result = reg128 {
+        f32: [
+            destination.f32[0] + destination.f32[1],
+            destination.f32[2] + destination.f32[3],
+            source.f32[0] + source.f32[1],
+            source.f32[2] + source.f32[3],
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:565]: https://www.felixcloutier.com/x86/movddup

File: src/rust/cpu/instructions_0f.rs [hunk 13]
```diff
@@ -1822,6 +1862,42 @@ pub unsafe fn instr_F30F59_mem(addr: i32, r: i32) {
...
+#[no_mangle]
+pub unsafe fn instr_660F7D(source: reg128, r: i32) {
+    // hsupbd xmm, xmm/mem128
+    let destination = read_xmm128s(r);
+    let result = reg128 {
+        f32: [
+            destination.f32[0] + destination.f32[1],
+            destination.f32[2] + destination.f32[3],
+            source.f32[0] + source.f32[1],
+            source.f32[2] + source.f32[3],
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:1891]: https://www.felixcloutier.com/x86/hsubpd

**Review by copy — COMMENTED** [review]
> sorry for being inpatient, but can you merge this pull request, since we have sse3 support?

I cleared the cache, because the nasmtests needed to be generated, and it found a couple of bugs.

**Comment by copy** [comment]
> Agreed, in CPU-Z detects SSE3, but in Linux's /proc/cpuinfo pni (Prescott New Instructions, second name of SSE3) is not detected

`pni` is set when using this PR (I wasn't aware of the alternative name, thanks!). So the cpuid part seems to work fine.

Screenshot_2023-09-11_16-42-28

**Comment by Ryan8085309853409** [comment]
Ok, will work on this when I'm home

**Comment by spetterman66** [comment]
(or maybe at school lmfao)

**Comment by BelleNottelling** [comment]
RIP to anyone who follows this repo and didn't unfollow this PR when it was originally created. (That's 161 emails from this PR)
image

**Comment by Ryan8085309853409** [comment]
well then you better do it huh

**Comment by spetterman66** [comment]
AYYYYYYYY IT PASSED

**Comment by spetterman66** [comment]
now it's ready to merge... right?

File: src/rust/cpu/instructions_0f.rs [hunk 14]
```diff
@@ -658,10 +681,23 @@ pub unsafe fn instr_660F16_mem(addr: i32, r: i32) {
...
-
+pub unsafe fn instr_F30F16(source: reg128, r: i32) {
+    // movshdup xmm1, xmm2/m128
+    let destination = read_xmm128s(r);
+    let result = reg128 {
+         f32: [
+            destination.f32[0] + source.f32[1],
+            destination.f32[1] + source.f32[1],
+            destination.f32[2] + source.f32[3],
+            destination.f32[3] + source.f32[3],
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:692]: This is a move instruction, it shouldn't add.

File: src/rust/cpu/instructions_0f.rs [hunk 15]
```diff
@@ -1822,6 +1858,40 @@ pub unsafe fn instr_F30F59_mem(addr: i32, r: i32) {
...
 #[no_mangle]
+pub unsafe fn instr_660F7C(source: reg128, r: i32) {
+    // haddpd xmm1, xmm2/m128
+    let destination = read_xmm128s(r);
+    let result = reg128 {
+        f32: [
+            destination.f32[0] * destination.f32[1],
+            destination.f32[2] * destination.f32[3],
+            source.f32[0] * source.f32[1],
+            source.f32[2] * source.f32[3],
```
Comment (copy) [inline src/rust/cpu/instructions_0f.rs:1869]: This is an add instruction, it shouldn't multiply.

**Comment by spetterman66** [comment]
do we need to clean github actions cache or something?

**Comment by copy** [comment]
> do we need to clean github actions cache or something?

I don't think it's a cache problem. Can you try `./tests/nasm/create_tests.js && make nasmtests` locally?

**Comment by spetterman66** [comment]
6 errors :)

**Comment by spetterman66** [comment]
***when will this be merged***

**Comment by copy** [comment]
Merged, cheers.

**Comment by cyberplant** [comment]
Well done, congratulations @spetterman66 !

**Comment by Ryan8085309853409** [comment]
You would have never thought a web dev could implement sse3 in an emulator :D

**Pull request merged**